### ISSUE TITLE
TypeScript: types vs. interfaces

### ISSUE THREAD

**Issue by jhnns** [peerigon/eslint-config-peerigon#64]
TypeScript has two ways to describe the shape of an object: a type alias and an interface. I would like to collect some arguments for the discussion around types vs interfaces.

Historically, interfaces were more powerful which is why a lot of people went with interfaces. But nowadays the differences are not that big anymore. To summarize:

- ~~Type aliases don't create new names in error messages~~ (not correct anymore)
- ~~Type aliases cannot be implemented by classes~~ (not correct anymore, only union types are not implementable which are impossible to do with interfaces anyway)
- Two type aliases with the same name throw an error whereas two interfaces with the same name are merged into a single one (known as "Declaration merging"). This is an advantage of interfaces when dealing with third-party libraries, but can be a disadvantage if there is a name conflict in the current file.
- In VSCode, interfaces are decorated with an "implemented by" hint and some type hints are cleaner

This blog post goes more into detail: https://medium.com/@martin_hotell/interface-vs-type-alias-in-typescript-2-7-2a8f1777af4c (also read the conclusion :))

Personally I prefer type aliases because of the following reasons:

### 1. It's less to write (and less to read)

```ts
type A = AnotherType & ADifferentType & {
    a: boolean;
};

// vs

interface A extends AnotherType, ADifferentType {
    a: boolean;
}
```

Ok, that's not a big difference, but look at the next example:

```ts
const Component: React.SFC<Readonly<{
    a: boolean;
    b: string;
    c: number;
}>> = ({a, b, c}) => {};

// vs

interface ComponentProps {
   readonly a: boolean;
   readonly b: string;
   readonly c: number;
}

const Component: React.SFC<ComponentProps> = ({a, b, c}) => {};
```

### 2. It's a single way to handle types

One problem I see is that you can't really use interfaces in an ergonomic way when you want to use [TS' utility types](https://www.typescriptlang.org/docs/handbook/utility-types.html), like `Readonly`:

```ts
// we have to create a mutable interface first
// before we can create a readonly type :(
interface A {}
type B = Readonly<A>; 
```

`type` on the other hand provides a single way to define and merge types as you like. For instance, union types are impossible with interfaces, but straightforward with type aliases.

We used type aliases in a big project and never really missed interfaces.

---

For me, there are only three valid reasons for interfaces:

### Nicer hints in VSCode

Maybe they will improve it for `type` as well. I don't see a reason why this is only implemented for interfaces.

### Third-party libs

It's often recommended to use interfaces for third-party libs because of declaration merging. This makes it possible to extend the type with own properties.

While this is true for libraries like Express that provide a single `Request` or `Response` object which is extended by middlewares, I don't think that it's true for third-party libraries that don't support this kind of API. Personally I even think that it's bad practice to have an API where objects can be extended by everyone. So why should I use interfaces if I don't want to support this kind of usage anyway.

### OO programming style

If your project/team heavily uses an object-oriented programming style interfaces make much more sense.

---

What do you think?

**Comment by jhnns** [comment]
Related discussion in the TypeScript ESLint repo: https://github.com/typescript-eslint/typescript-eslint/issues/142

---

Btw: I just found out that the VSCode Lens can also show references of types:

![Screenshot 2019-07-18 at 13 32 17](https://user-images.githubusercontent.com/781746/61454262-b4654100-a960-11e9-8060-32d729ba153c.jpg)
![Screenshot 2019-07-18 at 13 32 05](https://user-images.githubusercontent.com/781746/61454265-b6c79b00-a960-11e9-9105-517825a448ec.jpg)

Related discussion in the VSCode repo: https://github.com/microsoft/vscode/issues/76706

**Comment by jhnns** [comment]
I found an interesting use case for declaration merging I haven't considered yet:

```ts
const Todo = (name: string) => ({
    name,
    description: undefined as string | undefined,
});

interface Todo extends ReturnType<typeof Todo> {}

function renderTodo(todo: Todo) {
    // ...
}

renderTodo(Todo("Do something"));
```

([playground](http://www.typescriptlang.org/play/index.html#code/MYewdgzgLgBAKiAJiGBeGAKMBDAtgUwC4ZoAnASzAHMBKNAPkwG8AoGdmHAgGjY8XwRgFAA5Ry4YgFcwAgGaV8iGNggkoFajAA+MGfMWJeAXxoBuFi0pR8pOdmD54SFPgAeN2WoBK+KFNIwOABPEXwAHihQ-BA5Z2RGJmNLORlgcXAYUnxZWwRkDCgXYnyQOlYOGAB6KpgAOgaWZJZs3NJSjA6AIgARFAgQAigAC0oqLppzIA))

Here the `Todo` function is merged with the interface. With types you can't merge the `Todo` type with the `Todo` function:

```ts
const createTodo = (name: string) => ({
    name,
    description: undefined as string | undefined,
});

type Todo = ReturnType<typeof createTodo>;

function renderTodo(todo: Todo) {
    // ...
}

renderTodo(createTodo("Do something"));
```

([playground](http://www.typescriptlang.org/play/index.html#code/MYewdgzgLgBMBOBTAhlRAVEATEMC8MAFGMgLaIBcM08AlmAOYCU+AfEQN4BQMvMJ5ADQ8+WRBAS0ADlFrgqAVzBiAZvURYYyCNSh1GMAD4wlq9VmEBfJgG4uXKAE8piGJhz4YAJURQF8MHRnRAAeJxcQFTgkVAxsEFY7LhUlYFlwGCRlRHh3EEIoeKo8lm4+GAB6CpgAOjquS3sssVz4wgQUNDzCACIAEVwIEHIoAAt6Bh6mWyA))

However, you could still use classes:

```ts
class Todo {
    description?: string = undefined;
    constructor(
        public name: string,
    ) {}
}

function renderTodo(todo: Todo) {
    // ...
}

renderTodo(new Todo("Do something"));
```

([playground](http://www.typescriptlang.org/play/index.html#code/MYGwhgzhAEAqD2ATe0DeAoaXqIKYWACcBLABwBdj4A7AfgC5oJyTqBzaAXmgFdq8AZsWq5EAbkzZgNZoR7By8QgApJ2bKR4AjEMWDRqYALa5Gs4WwA0a6AEo0AX3RP0AvgqrVohXP1yEEZGVFZEZA+HsMdQB6aOgAOkTndHQfPwCkeGURAHc4TOUAIgARFAh4E3IACwtC21sxIA))

I don't think that this is relevant to us since creating objects with a capitalized regular function (like `Todo("Do something")`) is not very typical in JS land and our linting rules would complain about it anyway. But I wanted to add that use case since I haven't considered it yet.

**Comment by hpohlmeyer** [comment]
Two things I like about interfaces, that are problematic with types:

__Code lens visibility in VS Code__
Since types are aliases they show their definition instead of their name.
If you hover over the variables in [this example](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgILIN4Chm+TAewIC5kBnMKUAcwG4c8AjOKUkAVwFtHp6BfLFlCRYiFACFkEAB6QQAEzLIACsAQBrADyoANMgDkhAvoB8mAQgIgKyRqUkBeTPiKl9kCvuR96WMAE8ABxQAYWQnVQ1tPUMiU3pLazBkBFIwpwwXEgMPMC8fIA) you will see the name of the interface and the definition of the type. I think if the interface has a descriptive name it is way easier to understand what is behind it.

__Impossible types with intersecion types__
```ts
type X = {
    foo: number,
    bar: string
}

type Y = {
    foo: string,
    baz: boolean
}

type XY = X & Y;

const xy: XY = {
    foo: 3, // Impossible to get right, because it has to be a number and a string at the same time and won’t be number | string
    bar: 'test',
    baz: true
}
```

I think the result of `X & Y` is not clear immediately and can be very problematic if you use generics and do not know which types will be merged. Extending interfaces is far more predictable.

**Comment by jhnns** [comment]
> Code lens visibility in VS Code

Yes, I agree. Code lens is still better for interfaces. But sometimes it's even an advantage when you can inspect the type by just hovering it.

Since `type` is a type alias, I don't get why they don't just show the aliased name, but the whole underlying type. Isn't the purpose of an alias to use it "in place" of something else?

> Impossible types with intersection types

I agree. When using interfaces, the error message appears at the location where the actual error happened:

```ts
interface X {
    foo: number,
    bar: string
}

interface Y {
    foo: string,
    baz: boolean
}

interface XY extends X, Y {} // Interface 'XY' cannot simultaneously extend types 'X' and 'Y'. Named property 'foo' of types 'X' and 'Y' are not identical.
```

[Playground](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgBrIN4Chm+TAewIC5kQBXAWwCNoAaHPauKUgZzClAHMsBfLFlCRYiFAE1MjXIRLIOXENwZ5kzAF6lqRADYQ4IfoOHR4SNJIgAPSCAAmbNHWSSMArAgIgOyKwE9SVEkAXilVWVIAZmcAehjkAElKAAcCNjZgaj1kMAJkbggwZC5uAAswZ1oEOHI2FGAi0rhHXLUUODIqWihkAzte+U4eXqKwUpQ2OEoUMGBp3vtkAHcvQEwCItpOmmhkAB9BxV5VZlZkAHJIDjOVJjhNHKhyCH4gA)

I haven't experienced these kind of errors yet, but it's a valid point against types.

Here's an explanation why TypeScript can't detect impossible types: https://stackoverflow.com/a/39554027

**Comment by jhnns** [comment]
> Impossible types with intersection types

You could also come up with examples where this behavior of intersection types is an advantage:

```ts
type TypeA = {
    name: string;
};

type TypeB = {
    name: "Something";
};

type TypeAB = TypeA & TypeB;

const typeAB1: TypeAB = {
    // Works
    name: "Something",
};
const typeAB2: TypeAB = {
    // Type '"Something else"' is not assignable to type '"Something"'
    name: "Something else",
};

interface InterfaceA {
    name: string;
}

interface InterfaceB {
    name: "Something";
}

// Named property 'name' of types 'InterfaceA' and 'InterfaceB' are not identical.(2320)
interface InterfaceAB extends InterfaceA, InterfaceB {}
```

[Playground](http://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=28&pc=56#code/C4TwDgpgBAKuEEEoF4oG8BQVtQHYEMBbCALigGdgAnAS1wHMBuDAX2Y1ElngCEV0sOAsTIAiAMoB7YsAAWdeqOZsMHeN0gI+qOJqgAyDRB7sAxpNyUonRDwCMZXbf6YcUAPTuoAdUlUA1uSC2MKkUBLSEHIKogA0rMzmlsDW8FoATI5p2gJunkZQAOQRMvIMUBAANuQQooVQNOR4kin45OQ09AQARpXQwJKpXMVSpTGFwXhEYSVRZfQV1bXxKhh0wBBUAGb4ptAAkrgb27uIuULTZJS0DMqq65s7e1CHx0-G5yGX4aNzMXcYfIAOWmABMoGAqJJIFRQEVQvVJFshhAmoVXo9Tgh6vhcOD0UdMXseDiqNBcC0GqCIEcaKZ8JUAHQACnSAGZ0gAGACUa0JJ2eGIFzggAA8NnimkL3ghYi9+e8+GgWEA)

**Comment by jhnns** [comment]
Interfaces used to have the advantage that they can express recursive types via `this`. Since TypeScript 3.7 [recursive types are supported](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#more-recursive-type-aliases).

This is no error anymore:

```ts
type B = {
    a: boolean;
    b: B["a"];
};
```

[Playground](http://www.typescriptlang.org/play/?ssl=9&ssc=3&pln=6&pc=1#code/JYOwLgpgTgZghgYwgAgILIN4Chm+XALmQCMB7UgGwjhAG4c9iiwALYAZwG0AiObgXXoBfLFjABPAA4oAQsgC8mBrkIlyVGvTwkiMnn0FYhtIA)

**Comment by jhnns** [comment]
Btw: Here is a (closed) issue at the TypeScript repo about the CodeLens issues of types: https://github.com/microsoft/TypeScript/issues/13095

**Comment by jhnns** [comment]
The fact that interfaces can use the polymorphic `this` type (and types cannot) leads to a situation where interfaces have a clear advantage over types when using classes:

```ts
interface I {
    returnThis: () => this;
}

type T = {
    returnThis: () => T;
};

class ClassI implements I {
    // Error which is correct, because returnThis should return the instance
    returnThis = () => new ClassI();
}

class ClassT implements T {
    // No error :(
    returnThis = () => new ClassT();
}
```

[Playground](http://www.typescriptlang.org/play/?ssl=18&ssc=1&pln=1&pc=1#code/JYOwLgpgTgZghgYwgAgJLIN4Chm+VCMAVyhABUALYAZwC5kAKASmQF4A+ZMK6gbiwC+WLGACeABxRk2mHHgLFSlGvWZtOZfgP5YEAGzjVqyAMIGj6YAFtxeiFYjhj6bHmQB6d8gCiUKAHsoZAB3KgQKZBpkBECCBDAAGmQAIwgEOCJqFAUSch5kagp-Ij0AE3xCXK4KFFBqMDgQJDlcHKV81kYWDmQQCGDTc2pUZi1hfUNjM0npa1t7RzBjaVc8T2QAOX9kaACg2gYWisU8qM61Hr6B6aMyUcEsIA)

This leads me to the conclusion that interfaces should be used when you're describing objects that participate in an inheritance chain. Essentially: Objects created with `new Something()` and also objects that create instances (= Classes) should be described via interfaces.

For regular data objects (like React `props` for instance), where `this` has no meaning or is not used, you should use `type`. In short: If you used an object literal to create that thing, it should be a `type`.

I wonder whether that can be expressed as ESLint rule 😁
